From eab2dd8737e9b4bdfbb124db316d2f8a8b1cbe16 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:06:29 -0400
Subject: [usb] storage: Don't bind to uas devices if the uas driver is enabled

Message-id: <1410887463-20674-318-git-send-email-dzickus@redhat.com>
Patchwork-id: 95359
O-Subject: [RHEL7 PATCH 317/591] usb-storage: Don't bind to uas devices if the uas driver is enabled
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 5bfd5b5d8b9cd9e8ba1709f2b9dc35bd4b26c8b1
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Mon Oct 21 09:40:48 2013 +0100

    usb-storage: Don't bind to uas devices if the uas driver is enabled

    uas devices have 2 alternative settings on their usb-storage interface,
    one for usb-storage and one for uas. Using the uas driver is preferred, so if
    the uas driver is enabled, and the device has an uas alt setting, don't bind.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/storage/unusual_devs.h b/drivers/usb/storage/unusual_devs.h
index ad06255..00c1dda 100644
--- a/drivers/usb/storage/unusual_devs.h
+++ b/drivers/usb/storage/unusual_devs.h
@@ -2079,6 +2079,11 @@ UNUSUAL_DEV( 0xed10, 0x7636, 0x0001, 0x0001,
   "Digital MP3 Audio Player",
   USB_SC_DEVICE, USB_PR_DEVICE, NULL, US_FL_NOT_LOCKABLE ),
 
+/* Unusual uas devices */
+#if IS_ENABLED(CONFIG_USB_UAS)
+#include "unusual_uas.h"
+#endif
+
 /* Control/Bulk transport for all SubClass values */
 USUAL_DEV(USB_SC_RBC, USB_PR_CB),
 USUAL_DEV(USB_SC_8020, USB_PR_CB),
diff --git a/drivers/usb/storage/usb.c b/drivers/usb/storage/usb.c
index 1c0b89f..388f567 100644
--- a/drivers/usb/storage/usb.c
+++ b/drivers/usb/storage/usb.c
@@ -72,6 +72,10 @@
 #include "sierra_ms.h"
 #include "option_ms.h"
 
+#if IS_ENABLED(CONFIG_USB_UAS)
+#include "uas-detect.h"
+#endif
+
 /* Some informational data */
 MODULE_AUTHOR("Matthew Dharm <mdharm-usb@one-eyed-alien.net>");
 MODULE_DESCRIPTION("USB Mass Storage driver for Linux");
@@ -1035,6 +1039,12 @@ static int storage_probe(struct usb_interface *intf,
  int result;
  int size;
 
+ /* If uas is enabled and this device can do uas then ignore it. */
+#if IS_ENABLED(CONFIG_USB_UAS)
+ if (uas_use_uas_driver(intf, id))
+  return -ENXIO;
+#endif
+
  /*
   * If the device isn't standard (is handled by a subdriver
   * module) then don't accept it.
-- 
1.7.1